### ISSUE TITLE
TypeError: 'NoneType' object is not iterable

### ISSUE THREAD

**Issue by ursschutz** [papis/papis#182]
Trying to get papis working under windows 10:

```
C:\Users\scu>python --version
Python 3.7.3
C:\Users\scu>papis --version
papis, version 0.8.2
```

```
C:\Users\scu>papis -l sandra -v open

|Propagation of a Surface Wave Sustaining a Plasma Column at Atmospheric Pressure
| Moisan, M. and Pantel, R. and Hubert, J.
|  (1990) [Drude Paper]
 Three-Dimensional Fluid Model for Atmospheric Pressure Dielectric Barrier Discharge Plasma
  Muhammad M. Iqbal, Miles M. Turner
   (2015) [DBD plasma, atmospheric pressure, fluid model, He helium, N2 nitrogen]
 Nitrogen-Rich Plasma-Polymer Films for Biomedical Applications
  Florina Truica-Marasescu, Michael R. Wertheimer
   (2008) [NH3, ammonia, C2H4, ethylene, FTIR spectroscopy, spectra]
```

Opening the third document by "down-arrow" and "enter" is working as expected.
But when I try to open by keyword the command window shows this:

```
C:\Users\scu>papis -l sandra -v open Nitrogen
421-DEBUG:db:cache:Querying
421-DEBUG:db:cache:Getting documents from cache in C:\Users\scu\.cache\papis\database\3c8460a9ddf327f9da8dbd27caa10a20-papis
421-DEBUG:db:cache:Loaded documents (3 documents)
437-DEBUG:query_parser:Parsing search
437-DEBUG:query_parser:Parsed query = [['Nitrogen']]
562-DEBUG:filter:Filtering 3 docs (search Nitrogen) using 8 cores
562-DEBUG:filter:pool started
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\site-packages\papis\docmatcher.py", line 62, in return_if_match
    for parsed in cls.parsed_search:
TypeError: 'NoneType' object is not iterable
"""
```

The above exception was the direct cause of the following exception:

```
Traceback (most recent call last):
  File "C:\Users\scu\AppData\Local\Programs\Python\Python37-32\Scripts\papis-script.py", line 11, in <module>
    load_entry_point('papis==0.8.2', 'console_scripts', 'papis')()
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\site-packages\click\core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\site-packages\click\core.py", line 717, in main
    rv = self.invoke(ctx)
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\site-packages\click\core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\site-packages\click\core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\site-packages\click\core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\site-packages\papis\commands\open.py", line 170, in cli
    documents = papis.database.get().query(query)
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\site-packages\papis\database\cache.py", line 243, in query
    return filter_documents(docs, query_string)
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\site-packages\papis\database\cache.py", line 94, in filter_documents
    papis.docmatcher.DocMatcher.return_if_match, documents
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\multiprocessing\pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "c:\users\scu\appdata\local\programs\python\python37-32\lib\multiprocessing\pool.py", line 657, in get
    raise self._value
TypeError: 'NoneType' object is not iterable
```

**Comment by alejandrogallo** [comment]
Hi,

thank you for this posting.
This does not look nice... I'm trying to reproduce it.

In the mean time if you've got time, could you try to do this
with version v0.9 of papis? This is in the branch `v0.9`, so
you can install it as

```
pip install git+https://github.com/papis/papis.git@v0.9#egg=papis
```

Maybe you can also use a [virtual environment](https://virtualenv.pypa.io/en/stable/),
then you can install papis locally somewhere, for instance
to install the virtual environment just do

```
pip install virtualenv
```

and you can create a folder `urs` with a local python, for instance

```
virtualenv urs
./urs/bin/pip install git+https://github.com/papis/papis.git@v0.9#egg=papis
```

and you can use the papis in that folder then
```
./urs/bin/papis -l sandra open Nitrogen
```

this way you don't conflict with the papis installed already in your system.

I'm taking a look at this issue further, however the issue is in the database
of version 0.8, since in version 0.9 several things have changed in the database,
you might have it solved there, and plus it would help me test version 0.9
on windows.

Thanks so much,
Ale

**Comment by ursschutz** [comment]
Hi Ale,
Thanks for looking into this. I will have a look at version v0.9 next week.
Thanks
Urs

**Comment by ursschutz** [comment]
Hi Ale,
Sadly, the same happens here with papis V0.9 from git.

### Workaround:
In the meantime I stumbled upon a workaround:
If I use _papis open_, and then start writing parts of a keyword or title or author (within papis), the selection mechanism (narrowing down to valid documents) is working as intended, and I get very fast access to the documents I'm looking for.

### Error:
Try to open a paper directly with a keyword ('Canon', in my case) is not working as expected:

_papis -v open Canon_
```
499-DEBUG:db:cache:Querying
499-DEBUG:db:cache:Getting documents from cache in C:\Users\IEUser\.cache\papis\database\8984178ccd2ecebff1595422441e0380-papers
499-DEBUG:db:cache:Loaded documents (5 documents)
515-DEBUG:query_parser:Parsing search
515-DEBUG:query_parser:Parsed query = [['Canon']]
1249-DEBUG:filter:Filtering 5 docs (search Canon) using 4 cores
1249-DEBUG:filter:pool started
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\multiprocessing\pool.py", line 44, in mapstar
    return list(map(*args))
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\papis\docmatcher.py", line 62, in return_if_match
    for parsed in cls.parsed_search:
TypeError: 'NoneType' object is not iterable
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\IEUser\AppData\Local\Programs\Python\Python37\Scripts\papis-script.py", line 11, in <module>
    load_entry_point('papis==0.9', 'console_scripts', 'papis')()
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\click\core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\click\core.py", line 717, in main
    rv = self.invoke(ctx)
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\click\core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\click\core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\click\core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\papis\commands\open.py", line 170, in cli
    documents = papis.database.get().query(query)
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\papis\database\cache.py", line 243, in query
    return filter_documents(docs, query_string)
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\papis\database\cache.py", line 94, in filter_documents
    papis.docmatcher.DocMatcher.return_if_match, documents
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\multiprocessing\pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "c:\users\ieuser\appdata\local\programs\python\python37\lib\multiprocessing\pool.py", line 657, in get
    raise self._value
TypeError: 'NoneType' object is not iterable
```


### Steps to reproduce:

Install VirtualBox (as instructed by your preferred OS and installer.)

Download a free, limited virtual machine of windows10 ("Test for Microsoft Edge") for VirtualBox from here:
https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/
for example this one:
https://az792536.vo.msecnd.net/vms/VMBuild_20190311/VirtualBox/MSEdge/MSEdge.Win10.VirtualBox.zip

Install/import the virtual machine into virtualbox, and startup virtual win10 machine.

Within windows10:

Install python for windows:
Download Python for windows here:
https://www.python.org/downloads/windows/
For example this one:
https://www.python.org/downloads/release/python-373/
https://www.python.org/ftp/python/3.7.3/python-3.7.3-amd64.exe
Install python for windows, accepting to set the path during install.

Install git for windows from here:
https://gitforwindows.org/
Download for example this one:
https://github.com/git-for-windows/git/releases/download/v2.21.0.windows.1/Git-2.21.0-64-bit.exe
Install git for windows, accept default parameters.

open the command window (terminal):
_cmd_

upgrade pip to the latest version:
_python -m pip install --upgrade pip_

close and reopen the command window, check versions:
_cmd_

_python --version_
`Python 3.7.3`

_git --version_
`git version 2.22.0.windows.1`

Install papis from git:
_pip install git+https://github.com/papis/papis.git@v0.9#egg=papis_

Set HOME environment variable:
Edit the system environment variables. Search for _environment_
System Properties
       Advanced
       Environment Variables...
       New...
	  _HOME_
	  _C:\Users\IEUser_
	OK

Install Acrobat Reader DC

restart windows10

within windows, open cmd (the terminal)

_papis --version_
`papis, version 0.9`

configure papis:
_notepad .config/papis/config_

```
[papers]
dir = ~/Documents/papers

[settings]
default-library = papers
editor = notepad
opentool = 'C:\Program Files (x86)\Adobe\Acrobat Reader DC\Reader\AcroRd32.exe'
```

create ~/Documents/papers:
_mkdir Documents/papers_

add some papers... Edit some authors and titles if necessary.

Try to open a paper directly with a keyword ('Canon', in my case) is not working as expected:
_papis -v open Canon_

### Remark
I do not know a lot of python. In my opinion the problem lies in line 62 of file "c:\users\ieuser\appdata\local\programs\python\python37\lib\site-packages\papis\docmatcher.py".
The variable "cls.parsed_search" is not an enumerable datatype, and therefore the expression "for parsed in cls.parsed_search" is not working as intended. My python knowledge is insufficient to make an valid assumption how to repair this. Nor can I explain why this code is working in python for linux, but not in python for windows...

Regards
Urs

**Comment by alejandrogallo** [comment]
Hi Urs,

first of all I want to thank you for taking the time to help me
setting up a virtual machine. I just downloaded a windows 8 image
and tested it, and I can now reproduce the error, which means I
can start fixing it, I appreciate very much you taking time for
this.

It is indeed a problem of the query parser, which is weird, but I'm testing it.

I'll get to you later today or this week.

Thanks!

**Comment by wbrandenburger** [comment]
Servus,

I also use windows and need solutions for this platform due to some restrictions. I really appreciate your work and want to use papis in my studies as default literature management tool.

Yesterday I installed papis and ran in the same problem described above. I tried a simple solution and it works. Actually, it is a problem of the query parser but only in combination with the function of the multiprocessing package. The multiprocessing function map has to use the `return_if_match` function of class `DocMatcher` but it seems to initialize a new instance when conducting return_if_match in parallel. Hence. `cls.parsed_search` is not iterable because its `None`. A simple solution, with focus on simple (i am not familiar with python), is to define the `return_if_match` as a static method, which additionally takes a object containing the necessary variables of  `DocMatcher`-Object. In consequence, the multiprocessing function `map` takes besides the documents these variables, which can be solved with the related multiprocessing function `starmap`.

In docmatcher.py:

```Python
    @classmethod
    def get_parameter(cls):
        obj = {
            'parsed_search': cls.parsed_search,
            'matcher': cls.matcher,
            'doc_format': cls.doc_format
        }
        return obj

    @staticmethod
    def return_if_match_static(doc, obj):
        match = None
        for parsed in obj['parsed_search']:
            if len(parsed) == 1:
                search = parsed[0]
                sformat = None
            elif len(parsed) == 3:
                search = parsed[2]
                sformat = obj['doc_format.replace']('DOC_KEY', parsed[0])
            match = doc if obj['matcher'](doc, search, sformat) else None
            if not match:
                break
        return match
```
In cache.py:

``` Python
from itertools import repeat

    logger = logging.getLogger('filter')
    papis.docmatcher.DocMatcher.set_search(search)
    papis.docmatcher.DocMatcher.parse()
    papis.docmatcher.DocMatcher.set_matcher(match_document)
    np = multiprocessing.cpu_count()
    pool = multiprocessing.Pool(8)
    logger.debug(
        "Filtering {} docs (search {}) using {} cores".format(
            len(documents),
            search,
            np
        )
    )
    begin_t = 1000 * time.time()
    # CHANGES: get variables of docmatcher object and start multiprocessing with starmap
    obj = papis.docmatcher.DocMatcher.get_obj()
    result = pool.starmap(papis.docmatcher.DocMatcher.return_if_match_attempt, zip(documents, repeat(obj)))
    
    pool.close()
    pool.join()
    _delta = 1000 * time.time() - begin_t
    filtered_docs = [d for d in result if d is not None]
    logger.debug("done ({} ms) ({} docs)".format( _delta, len(filtered_docs)))
    return filtered_docs
```

This solution works fine, but its not really a stringent one, because it can be discussed whether the class DocMatcher is meaningful. Moreover, i am not familiar with python and there could be a solution to use the function `return_if_match` of `DocMatcher` with instance variables. 

Does it make sense, what I have written...?

**Comment by wbrandenburger** [comment]
I added the changes described above to my [fork of papis](https://github.com/wbrandenburger/papis).

I have tested papis with several settings and haven't encountered any major problems. Merely, the main problem on windows is to correctly interact with user's software for editing and opening files especially their specific marks.

**Comment by alejandrogallo** [comment]
Hello @wbrandenburger and @ursschutz 
Thank you very much both of you for all.
Now I'm back fully recovered and I've spent today some time taking a look at this.
While the solution of @wbrandenburger works and it's actually a good one, it changes
a lot of things and I'll actually review it to include some or all of it for the new version.
But since `v0.9` should have been out there long time ago, I fixed this with a quick fix today.

The problem, it seems, is that windows is not playing well with static parameters for classes,
and calling the `pool.map` method was simply parsing an empty query.
The quick fix I did is to shut down the multiprocessing on windows.
Therefore the main issue of this section is solved, however some work has to be done to further
improve and cure such problems in the future.
